lib/web_ui/lib/src/engine/bitmap_canvas.dart

```diff
@@ -436,8 +436,22 @@ class BitmapCanvas extends EngineCanvas {
       if (requiresClipping) {
         restore();
       }
-      _canvasPool.allocateExtraCanvas();
     }
+    _allocateNewCanvas();
+  }
+
+  // Should be called when we add new html elements into rootElement so that
+  // paint order is preserved.
+  //
+  // For example if we draw a path and then a paragraph and image:
+  //   - rootElement
+  //   |--- <canvas>
+  //   |--- <p>
+  //   |--- <img>
+  // Any drawing operations after these tags should allocate a new canvas,
+  // instead of drawing into earlier canvas.
+  void _allocateNewCanvas() {
+    _canvasPool.allocateExtraCanvas();
```

Comment:
Let's rename `_allocateNewCanvas` and `allocateExtraCanvas` to something like "closeCurrentCanvas", since they do not allocate a new canvas. Otherwise it can be confusing when looking at the profiles that `allocateExtraCanvas` takes 1 nanosecond, but `drawCircle` takes 10 milliseconds.

Reply:
Done. Thanks for suggestion.